### PR TITLE
Refactor ScanStore

### PR DESCRIPTION
**Pull request by redhoteggplant**
* Fix updating cartItems with useState hook
* Remove `shoppingItems` from code
* Remove `idxMap`

### CODE REVIEW

File: client/src/components/StoreList.tsx [hunk 1]
```diff
@@ -117,18 +123,73 @@ function StoreList() {
     }
   };
 
+  // Scan Store's QR code
+  const storeQR = async () => {
+    if (window.location == window.parent.location) {
```
Comment (BhavyaLight) [inline client/src/components/StoreList.tsx:128]: why is this needed?  Can add comment or extract to a function

File: client/src/components/StoreList.tsx [hunk 2]
```diff
@@ -117,18 +123,73 @@ function StoreList() {
...
+  };
+
+  const processImageBarcode = async (img: HTMLImageElement) => {
+    const codeReader = new BrowserMultiFormatReader();
+    if (img != null) {
+      const result = await codeReader
+        .decodeFromImage(img)
+        .then((res: any) => res)
+        .catch((err: any) => {
+          console.log(err);
```
Comment (BhavyaLight) [inline client/src/components/StoreList.tsx:152]: remove console.log. These are alright for dev but better not to check in.

File: client/src/components/StoreList.tsx [hunk 3]
```diff
@@ -117,18 +123,73 @@ function StoreList() {
...
+          return "";
+        });
+      if (result) {
+        window.location.href = "/store?" + result.text;
+      } else {
+        console.log("Unable to redirect, QR Code not scanned correctly");
```
Comment (BhavyaLight) [inline client/src/components/StoreList.tsx:158]: ditto

File: client/src/components/StoreList.tsx [hunk 4]
```diff
@@ -117,18 +123,73 @@ function StoreList() {
...
+      }
+    } else {
+      console.log("Cannot find element: storeQRImgSrc");
```
Comment (BhavyaLight) [inline client/src/components/StoreList.tsx:161]: ditto

File: client/src/components/StoreList.tsx [hunk 5]
```diff
@@ -117,18 +123,73 @@ function StoreList() {
...
+      ) as HTMLImageElement;
+      processImageBarcode(img);
+    } else {
+      const imgReq = {
+        allowedMimeTypes: ["image/jpeg"],
+        allowedSources: ["camera"], // Restrict to camera scanning only
+      };
+      const imgRes = await window.microapps
+        .requestMedia(imgReq)
+        .then((res: any) => res);
```
Comment (BhavyaLight) [inline client/src/components/StoreList.tsx:140]: ```.then(res: any)``` If you are using the result object as it is, then you dont need an additional  ```then```, it is essentially redundant.

File: client/src/components/StoreList.tsx [hunk 6]
```diff
@@ -117,18 +123,73 @@ function StoreList() {
...
+      setUploadImg(imgRes);
+    }
+  };
+
+  const processImageBarcode = async (img: HTMLImageElement) => {
+    const codeReader = new BrowserMultiFormatReader();
+    if (img != null) {
+      const result = await codeReader
+        .decodeFromImage(img)
+        .then((res: any) => res)
```
Comment (BhavyaLight) [inline client/src/components/StoreList.tsx:150]: ditto for then being redundant

File: client/src/components/StoreList.tsx [hunk 7]
```diff
@@ -117,18 +123,73 @@ function StoreList() {
...
+      const imgRes = await window.microapps
+        .requestMedia(imgReq)
+        .then((res: any) => res);
+      setUploadImg(imgRes);
+    }
+  };
+
+  const processImageBarcode = async (img: HTMLImageElement) => {
+    const codeReader = new BrowserMultiFormatReader();
+    if (img != null) {
```
Comment (BhavyaLight) [inline client/src/components/StoreList.tsx:147]: ```if (img) ``` should suffice

File: client/src/components/StoreList.tsx [hunk 8]
```diff
@@ -117,18 +123,73 @@ function StoreList() {
...
   return (
     <div className="StoreList">
+      {uploadImg.mimeType && (
+        <img
+          hidden={true}
+          id="storeQRImgSrc"
+          src={"data:" + uploadImg.mimeType + ";base64," + uploadImg.bytes}
+        />
+      )}
+      <img hidden={true} id="sampleStoreQRImgSrc" src={SampleStoreQR} />
```
Comment (BhavyaLight) [inline client/src/components/StoreList.tsx:181]: why does it hide the img?

File: client/src/pages/Login.tsx [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+import Logo from "./../img/Logo.png";
+import "./../css/Login.css";
+
+// Applease typescript
+declare const window: any;
+// Grab handle to our microapps js library
+const microapps = window.microapps;
+
+function Login() {
+  const is_web = window.location == window.parent.location;
```
Comment (BhavyaLight) [inline client/src/pages/Login.tsx:14]: (Please file a new issue and leave TODO )
if this check is being done in multiple places, then you should extract it out to a common function!
Also, it might be better to check if window is inside iFrame instead of comparing with the urls e.g.
```
 window.self !== window.top;
```

File: client/src/components/StoreList.tsx [hunk 10]
```diff
@@ -117,18 +123,73 @@ function StoreList() {
...
+      } else {
+        console.log("Unable to redirect, QR Code not scanned correctly");
+      }
+    } else {
+      console.log("Cannot find element: storeQRImgSrc");
+    }
+  };
+
+  useEffect(() => {
+    if (uploadImg.mimeType) {
```
Comment (BhavyaLight) [inline client/src/components/StoreList.tsx:166]: should ``` if (uploadImg)``` itself be alright? is there a special reason to check ```uploadImg.mimeType``` ?

File: client/src/components/StoreList.tsx [hunk 11]
```diff
@@ -117,18 +123,73 @@ function StoreList() {
...
+      const img = document.getElementById("storeQRImgSrc") as HTMLImageElement;
+      processImageBarcode(img);
+    }
+  }, [uploadImg]);
+
   return (
     <div className="StoreList">
+      {uploadImg.mimeType && (
```
Comment (BhavyaLight) [inline client/src/components/StoreList.tsx:174]: I think the image upload logic is certainly complex and should be extracted out to a smaller react component and called inside the StoreList.

File: client/src/pages/Login.tsx (hunk 9 above)
Comment (devYaoYH) [reply client/src/pages/Login.tsx:14]: resolved in #22

File: client/src/components/StoreList.tsx (hunk 11 above)
Comment (devYaoYH) [reply client/src/components/StoreList.tsx:174]: TODO #66

File: client/src/components/StoreList.tsx (hunk 10 above)
Comment (devYaoYH) [reply client/src/components/StoreList.tsx:166]: uploadImg is initialized to a non-null value, but mimeType indicates whether we have a valid image type to process

File: client/src/components/StoreList.tsx (hunk 8 above)
Comment (devYaoYH) [reply client/src/components/StoreList.tsx:181]: sample QR code, we do not want this displayed to users

File: client/src/components/StoreList.tsx (hunk 7 above)
Comment (devYaoYH) [reply client/src/components/StoreList.tsx:147]: resolved in #22

File: client/src/components/StoreList.tsx (hunk 5 above)
Comment (devYaoYH) [reply client/src/components/StoreList.tsx:140]: resolved in #22

File: client/src/components/StoreList.tsx (hunk 2 above)
Comment (devYaoYH) [reply client/src/components/StoreList.tsx:152]: resolved in #22

File: client/src/components/StoreList.tsx (hunk 1 above)
Comment (devYaoYH) [reply client/src/components/StoreList.tsx:128]: resolved in #22

File: client/src/components/StoreList.tsx [hunk 12]
```diff
@@ -21,6 +25,8 @@ function StoreList() {
   const [identity, setIdentity] = useState<IdentityToken>(emptyIdentityToken);
   const [nearbyPlaces, setNearbyPlaces] = useState<GMapPlace[]>([]);
   const [isLoading, setIsLoading] = useState(false);
+  // Uploaded image data
```
[the remark this answers is not in the document]
Comment (devYaoYH) [reply client/src/components/StoreList.tsx:28]: resolved in #22

File: client/src/components/StoreList.tsx (hunk 8 above)
Comment (BhavyaLight) [reply client/src/components/StoreList.tsx]: You mean like a test QR code?

File: client/src/pages/ScanStore.tsx [hunk 13]
```diff
@@ -19,152 +19,94 @@ import AddIcon from "@material-ui/icons/Add";
...
-      cartItems.splice(idx, 1);
-    } else {
-      cartItems[idx].quantity = quantity;
+    const [item]: Item[] = await fetchJson(data, ITEMS_API);
+    if (item) {
+      const cartItem: CartItem = {
+        item: item,
+        quantity: 1,
+      };
+      updateCart((cartItems) => [...cartItems, cartItem]);
```
Comment (BhavyaLight) [inline client/src/pages/ScanStore.tsx:81]:
```suggestion
      updateCart([...cartItems, {
        item: item,
        quantity: 1,
      }]);
```

File: client/src/pages/ScanStore.tsx [hunk 14]
```diff
@@ -19,152 +19,94 @@ import AddIcon from "@material-ui/icons/Add";
...
-    setShoppingList(cartItems);
+    // TODO (#59): Notify user if barcode is invalid or item not found
   };
 
   const renderShoppingList = () => {
-    let UI_List = [];
-    for (let i = 0; i < shoppingList.length; ++i) {
-      UI_List.push(
-        <Grid key={"grid" + i} item xs={12}>
+    return cartItems.map((cartItem) => {
```
Comment (BhavyaLight) [inline client/src/pages/ScanStore.tsx:87]: Where is this called?

**Comment by BhavyaLight** [comment]
Merge after resolving comments!

File: client/src/pages/ScanStore.tsx (hunk 14 above)
Comment (redhoteggplant) [reply client/src/pages/ScanStore.tsx:87]: In line 160 (hidden)

File: client/src/components/StoreList.tsx (hunk 8 above)
Comment (devYaoYH) [reply client/src/components/StoreList.tsx]: yes, that is correct, only runs when we detect we're not in the microapps environment
